### ISSUE TITLE
Question: Missing EC2 instances (because of Elastic Beanstalk?)

### ISSUE THREAD

**Issue by yannvanhalewyn** [duo-labs/cloudmapper#131]
Hey guys!

Thanks a bunch for making and open sourcing this tool!

I'm just starting out with it, so I might've overlooked something. Basically I don't see my EC2 instances behind the load balancers. Here's a subset showing the relevant pieces:

![cloudmapper 1](https://user-images.githubusercontent.com/11170603/43413529-93963402-9430-11e8-87cd-aa285c454bf9.png)

There are multiple EB applications in the VPC and they all appear like this. The only EC2 instance I can see is the Bastion server.

Is this the expected behavior, and if not what should I look for when debugging it? Some extra info:

- The instances SG only accepts http (80) requests from the load balancer
- The instances are on a private subnet which routes outgoing traffic to a NAT Gateway
- The NAT is also not rendered (expected behavior?)

Not sure how much more information you might need to help me, please let me know.

Am I missing something over the collect / prepare step? Or is this the expected result?

**Comment by yannvanhalewyn** [comment]
Some more information: ec2-describe-instances returns the correct data, and `cloudmapper stats` also confirms the expected amount of ec2 instances.

**Comment by 0xdabbad00** [comment]
If the NAT is AWS infrastructure (and not an EC2 you are running) then it will not be displayed.  The ElasticBeanstalk EC2's should be displayed though.  From the dark black line around the Public node, it tells me you deleted the some of the nodes from the image.  Is it possible you deleted the ElasticBeanstalk nodes?  They should show up just like any other EC2.  Another possibility is if you performed some filtering in your `prepare` step that filtered out these nodes.

**Comment by yannvanhalewyn** [comment]
@0xdabbad00 Thanks for the reply!

> If the NAT is AWS infrastructure (and not an EC2 you are running) then it will not be displayed.

Yes, it's an AWS managed NAT gateway. And it's fine not to see it then.

> From the dark black line around the Public node, it tells me you deleted the some of the nodes from the image. Is it possible you deleted the ElasticBeanstalk nodes?

I deleted the rest of our infrastructure in order not to just dump it all publicly. But no, I didn't delete anything that relates to this app in particular. I deleted DB's, ASG's and ELB's for other applications.

> They should show up just like any other EC2. Another possibility is if you performed some filtering in your prepare step that filtered out these nodes.

I didn't run anything special. Just the bare commands. To be sure we're on the same page:

```
python cloudmapper.py collect --account cloudmapper-bm
python cloudmapper.py prepare
python cloudmapper.py webserver
```

I did the same thing with my personal account which has only 1 simple elastic beanstalk app, and I only see the ASG over there (there are no ELB's and I don't see the EC2 instance either).

## My Machine

```
$ python --version
Python 2.7.15

$ system_profiler SPSoftwareDataType
Software:

    System Software Overview:

      System Version: macOS 10.13.5 (17F77)
      Kernel Version: Darwin 17.6.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: Yann’s 15" van Brightin
      User Name: Yann Vanhalewyn (yannvanhalewyn)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 5 days 6:29
```

**Comment by 0xdabbad00** [comment]
Could you try putting some debug code in at this point: https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/commands/prepare.py#L238

After the call to `ec2 = Ec2...` add `print(ec2.name)` and ensure your EC2 is showing up there.   I'm trying to identify at which point this could be getting filtered.  If it is not showing up there, then above that `ec2 = Ec2...` line, add `print(json.dumps(ec2_json, indent=4))`.

**Comment by 0xdabbad00** [comment]
Also, CloudMapper has been upgraded to work with Python 3 now although I guess I am not enforcing that strongly.  I can't imagine how that would be an issue, but you could try updating for that.

**Comment by yannvanhalewyn** [comment]
@0xdabbad00 Thanks for the info, I'll try those things when I get the change 😄

**Comment by yannvanhalewyn** [comment]
@0xdabbad00 All my instances appear when I put the print statement where you suggested, so I'm guessing the collect and prepare steps are at fault and I can find the instances in the resulting `web/data.json`. I guess I'll go through a bit of the javascript to see if I can find something out.

**Comment by 0xdabbad00** [comment]
Weird.  If the instances are in the `web/data.json` it means something is broken in displaying it.  My javascript code is really just setting things up for cytoscape.js.  You can see here where the `data.json` is being loaded: https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/web/js/cloudmap.js#L39

Open the developer console in the browser and see if any errors are being displayed there.   
One possibility is check the `type` of the instance in the `data.json`.  Maybe for some reason that is being set to something that isn't in the `style.json`: https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/web/style.json#L150

**Comment by yannvanhalewyn** [comment]
@0xdabbad00 Thanks for the hint about `type`. I got a bit closer to the issue thanks to that. Every EC2 instance was tagged as type `autoscaling`. Going through the code I notice that this is actually kinda makes sense:

https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/shared/nodes.py#L287-L289

The instance is in an autoscaling group, so the instance is rendered as one. I was a bit confused only seing the autoscaling groups and not the instances.

I now have the issue that I only see instances in 1 of two AZs. I should see one in each private subnet but only see the first one. I do see that both instances take on the ID of the autoscaling group and that probably causes crytoscape to only render one.

Thanks for the help! I'm getting closer.

Update: indeed, changing:

https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/shared/nodes.py#L289

to take on the `json_blob["InstanceId"]` reveals all the instances as expected. Would you be open to a PR displaying all instances inside autoscaling groups?

**Comment by 0xdabbad00** [comment]
Ah, yes, in order to try to avoid having dozens or more EC2's that all look exactly the same due to being in an autoscaling group, I try to be smart and reduce that to a single node.  Unfortunately, the code basically just picks one EC2 within that group as the "winner" and uses that one.  If your autoscaling group extends across multiple AZ's, that means the group will only be represented as a node in one AZ.

Two changes are needed:
- Change 1: A flag would work to just display all of the EC2s in the autoscaling group.
- Change 2: Also, when the autoscaling groups are condensed to a single node, I should show separate autoscaling group nodes for each AZ.

I'll begin working on CloudMapper again tomorrow to fix up some issues, so I can add that flag tomorrow (Change 1).  Change 2 will be a little harder, but I'll see what I can do.  I'll happily merge in your PR though if you want to do something for this.

The new code for the command-line parsing would go here: https://github.com/duo-labs/cloudmapper/blob/3249aacb2d701847a5360e2d56f7243f44c909be/commands/prepare.py#L360
I think I'll use names `--condense-asgs` (default, which is how it currently works) and `--no-condense-asgs`.

**Comment by yannvanhalewyn** [comment]
The flags would be an easy start indeed! I think if setting the ASG's local_id's to be the ASG + subnet-id it's a quick way to condense multiple instances but still show them accross AZ's.

https://github.com/duo-labs/cloudmapper/pull/135

**Comment by 0xdabbad00** [comment]
Thank you! I merged in your code with some changes to ensure that collapsing ASGs remains the default, so you'll want to run prepare with `--no-collapse-asgs`

**Comment by 0xdabbad00** [comment]
I've created another issue for the fix needed here (#137), which is going to be low priority for me.

**Comment by yannvanhalewyn** [comment]
Great!
